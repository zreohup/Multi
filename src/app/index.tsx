import React from 'react'
import { View } from 'tamagui'
import { ActivityIndicator } from 'react-native'

/**
 * This is a dummy screen. Expo automatically renders it when it constructs the app.
 * If we don't have an index file it will pick whatever it sees fit. This is a placeholder.
 *
 * The actual navigation to either onboarding flow or a safe happens inside the NavigationGuardHOC
 *
 * @constructor
 */
function IndexScreen() {
  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <ActivityIndicator />
    </View>
  )
}

export default IndexScreen
