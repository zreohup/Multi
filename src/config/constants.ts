import Constants from 'expo-constants'
import { Platform } from 'react-native'

// export const isProduction = process.env.NODE_ENV === 'production'
// TODO: put it to get from process.env.NODE_ENV once we remove the mocks for the user account.
export const isProduction = process.env.EXPO_PUBLIC_APP_VARIANT === 'production'
export const isAndroid = Platform.OS === 'android'
export const isTestingEnv = process.env.NODE_ENV === 'test'
export const isStorybookEnv = Constants?.expoConfig?.extra?.storybookEnabled === 'true'
export const POLLING_INTERVAL = 15_000

export const COMING_SOON_MESSAGE = 'This feature is coming soon.'
export const COMING_SOON_TITLE = 'Coming soon'

export const GATEWAY_URL_PRODUCTION =
  process.env.EXPO_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
export const GATEWAY_URL_STAGING = process.env.EXPO_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
export const GATEWAY_URL = isProduction ? GATEWAY_URL_PRODUCTION : GATEWAY_URL_STAGING

export const SECURITY_CERTIFICATE_HASH_BASE64 = process.env.EXPO_PUBLIC_SECURITY_SERTIFICATE_HASH_BASE64
export const SECURITY_WATCHER_MAIL = process.env.EXPO_PUBLIC_SECURITY_WATCHER_MAIL
export const SECURITY_RASP_ENABLED = process.env.EXPO_PUBLIC_SECURITY_RASP_ENABLED === 'true'

/**
 * The version of the onboarding flow.
 * If we change it and need all users to see it again, we can bump the version here.
 */
export const ONBOARDING_VERSION = 'v1'

export const SAFE_WEB_URL = 'https://app.safe.global'
export const SAFE_WEB_TRANSACTIONS_URL = `${SAFE_WEB_URL}/transactions/tx?safe=:safeAddressWithChainPrefix&id=:txId`
export const SAFE_WEB_FEEDBACK_URL =
  'https://docs.google.com/forms/d/e/1FAIpQLSfJXkNNsZqVtg3w3dwk-YrTNutQ00n3MMfLtH-dN8zSHaJu5Q/viewform?usp=dialog'
