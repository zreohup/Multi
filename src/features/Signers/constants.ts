import { SignerSection } from './components/SignersList/SignersList'

export const groupedSigners: Record<string, SignerSection> = {
  imported: {
    id: 'imported_signers',
    title: 'My signers',
    data: [],
  },
  notImported: {
    id: 'not_imported_signers',
    title: 'Not imported signers',
    data: [],
  },
}
