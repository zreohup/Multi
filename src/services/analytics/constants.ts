import type { TransactionInfoType, TransferInfoType, SettingsInfoType } from './types'

export const ANALYTICS_LABELS = {
  BASE_TYPES: {
    Creation: 'safe_creation',
    Custom: 'custom',
    Transfer: 'transfer',
    SettingsChange: 'settings_change',
    SwapOrder: 'swap_order',
    SwapTransfer: 'swap_transfer',
    TwapOrder: 'twap_order',
    NativeStakingDeposit: 'native_staking_deposit',
    NativeStakingValidatorsExit: 'native_staking_exit',
    NativeStakingWithdraw: 'native_staking_withdraw',
    VaultDeposit: 'vault_deposit',
    VaultRedeem: 'vault_redeem',
    SwapAndBridge: 'swap_and_bridge',
    Swap: 'swap',
  } as const satisfies Record<TransactionInfoType, string>,

  TRANSFER_TYPES: {
    ERC20: 'transfer_token',
    ERC721: 'transfer_nft',
    NATIVE_COIN: 'transfer_native',
  } as const satisfies Record<TransferInfoType, string>,

  SETTINGS_TYPES: {
    ADD_OWNER: 'owner_add',
    REMOVE_OWNER: 'owner_remove',
    SWAP_OWNER: 'owner_swap',
    CHANGE_THRESHOLD: 'owner_threshold_change',
    DELETE_GUARD: 'guard_remove',
    DISABLE_MODULE: 'module_remove',
    ENABLE_MODULE: 'module_enable',
    SET_FALLBACK_HANDLER: 'fallback_handler_set',
    SET_GUARD: 'guard_set',
    CHANGE_MASTER_COPY: 'safe_update',
  } as const satisfies Record<SettingsInfoType, string>,

  ENHANCED: {
    batch_transfer_token: 'batch_transfer_token',
    batch: 'batch',
    rejection: 'rejection',
    typed_message: 'typed_message',
    safeapps: 'safeapps',
    walletconnect: 'walletconnect',
    activate_without_tx: 'activate_without_tx',
    activate_with_tx: 'activate_with_tx',
  } as const,
} as const
