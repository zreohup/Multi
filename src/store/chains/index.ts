import { apiSliceWithChainsConfig, chainsAdapter, initialState } from '@safe-global/store/gateway/chains'
import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '..'
import { Chain } from '@safe-global/store/gateway/AUTO_GENERATED/chains'
import { selectActiveSafe } from '../activeSafeSlice'

const selectChainsResult = apiSliceWithChainsConfig.endpoints.getChainsConfig.select()

const selectChainsData = createSelector(selectChainsResult, (result) => {
  return result.data ?? initialState
})

const { selectAll: selectAllChains, selectById } = chainsAdapter.getSelectors(selectChainsData)

export const selectChainById = (state: RootState, chainId: string) => selectById(state, chainId)

export const selectActiveChain = createSelector(
  [selectActiveSafe, (state: RootState) => state],
  (activeSafe, state) => {
    if (!activeSafe) {
      return null
    }
    return selectChainById(state, activeSafe.chainId)
  },
)

export const selectAllChainsIds = createSelector([selectAllChains], (chains: Chain[]) =>
  (chains || []).map((chain) => chain.chainId),
)

export const selectActiveChainCurrency = createSelector(
  [selectActiveChain, (state: RootState) => state],
  (activeChain) => {
    return activeChain?.nativeCurrency
  },
)

export const getChainsByIds = createSelector(
  [
    // Pass the root state and chainIds array as dependencies
    (state: RootState) => state,
    (_state: RootState, chainIds: string[]) => chainIds,
  ],
  (state, chainIds) => {
    return (chainIds || []).map((chainId) => selectById(state, chainId)).filter(Boolean)
  },
)

export const { useGetChainsConfigQuery } = apiSliceWithChainsConfig
export { selectAllChains }
