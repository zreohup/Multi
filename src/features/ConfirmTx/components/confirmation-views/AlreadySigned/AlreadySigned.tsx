import { SafeButton } from '@/src/components/SafeButton'
import { SafeFontIcon } from '@/src/components/SafeFontIcon'
import { SAFE_WEB_TRANSACTIONS_URL } from '@/src/config/constants'
import { selectChainById } from '@/src/store/chains'
import { useAppSelector } from '@/src/store/hooks'
import React, { useCallback } from 'react'
import { Linking, TouchableOpacity } from 'react-native'
import { Text, View, YStack } from 'tamagui'

type Props = {
  txId: string
  safeAddress: string
  chainId: string
  hasEnoughConfirmations: boolean
}

export function AlreadySigned({ txId, safeAddress, chainId, hasEnoughConfirmations }: Props) {
  const chain = useAppSelector((state) => selectChainById(state, chainId))
  const onPressGoToWebApp = useCallback(() => {
    const url = SAFE_WEB_TRANSACTIONS_URL.replace(
      ':safeAddressWithChainPrefix',
      `${chain?.shortName}:${safeAddress}`,
    ).replace(':txId', txId)

    Linking.openURL(url)
  }, [txId, safeAddress, chainId])

  return (
    <YStack justifyContent="center" gap="$4" alignItems="center" paddingHorizontal={'$4'}>
      <Text fontSize="$4" fontWeight={400} textAlign="center" color="$textSecondaryLight">
        {hasEnoughConfirmations
          ? 'This transaction can be executed in the web app only.'
          : 'Can be executed once the threshold is reached'}
      </Text>

      {hasEnoughConfirmations && (
        <>
          <TouchableOpacity onPress={onPressGoToWebApp}>
            <View flexDirection="row" alignItems="center" gap="$2">
              <Text fontSize="$4" fontWeight={700} textAlign="center" color="$color">
                Go to Web app
              </Text>
              <SafeFontIcon name="external-link" size={16} color="$color" />
            </View>
          </TouchableOpacity>

          <View height={50} width="100%">
            <SafeButton height="100%" rounded fullscreen fontWeight={600} disabled testID="confirm-button">
              Confirm
            </SafeButton>
          </View>
        </>
      )}
    </YStack>
  )
}
