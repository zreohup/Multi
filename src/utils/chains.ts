import { Chain } from '@safe-global/store/gateway/AUTO_GENERATED/chains'

export const getAvailableChainsNames = (chains: Chain[]) => {
  if (chains.length === 0) {
    return ''
  }

  if (chains.length === 1) {
    return chains[0].chainName
  }

  return (
    chains
      .slice(0, -1)
      .map((chain) => chain.chainName)
      .join(', ') +
    ' and ' +
    chains[chains.length - 1].chainName
  )
}
