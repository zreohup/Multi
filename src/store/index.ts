import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist'
import { reduxStorage } from './storage'
import txHistory from './txHistorySlice'
import activeSafe from './activeSafeSlice'
import activeSigner from './activeSignerSlice'
import signers from './signersSlice'
import delegates from './delegatesSlice'
import myAccounts from './myAccountsSlice'
import notifications from './notificationsSlice'
import addressBook from './addressBookSlice'
import settings from './settingsSlice'
import safes from './safesSlice'
import safeSubscriptions from './safeSubscriptionsSlice'
import biometrics from './biometricsSlice'
import { cgwClient, setBaseUrl } from '@safe-global/store/gateway/cgwClient'
import devToolsEnhancer from 'redux-devtools-expo-dev-plugin'
import { GATEWAY_URL, isTestingEnv } from '../config/constants'
import { web3API } from './signersBalance'
import { setBaseUrl as setSDKBaseURL } from '@safe-global/safe-gateway-typescript-sdk'
import { createFilter } from '@safe-global/store/utils/persistTransformFilter'
import { setupMobileCookieHandling } from './utils/cookieHandling'
import notificationsMiddleware from './middleware/notifications'
import analyticsMiddleware from './middleware/analytics'
import { setBackendStore } from '@/src/store/utils/singletonStore'

setSDKBaseURL(GATEWAY_URL)
setBaseUrl(GATEWAY_URL)

// Set up mobile-specific cookie handling
setupMobileCookieHandling()

const cgwClientFilter = createFilter(
  cgwClient.reducerPath,
  ['queries.getChainsConfig(undefined)', 'config'],
  ['queries.getChainsConfig(undefined)', 'config'],
)

const persistConfig = {
  key: 'root',
  version: 1,
  storage: reduxStorage,
  blacklist: [web3API.reducerPath, 'myAccounts'],
  transforms: [cgwClientFilter],
}

export const rootReducer = combineReducers({
  txHistory,
  safes,
  activeSigner,
  activeSafe,
  notifications,
  addressBook,
  myAccounts,
  signers,
  delegates,
  settings,
  safeSubscriptions,
  biometrics,
  [web3API.reducerPath]: web3API.reducer,
  [cgwClient.reducerPath]: cgwClient.reducer,
})

// Define the type for the root reducer
export type RootReducerState = ReturnType<typeof rootReducer>

// Use the persistReducer with the correct types
const persistedReducer = persistReducer<RootReducerState>(persistConfig, rootReducer)

export const makeStore = () =>
  configureStore({
    reducer: persistedReducer,
    devTools: false,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        },
      }).concat(cgwClient.middleware, web3API.middleware, notificationsMiddleware, analyticsMiddleware),
    enhancers: (getDefaultEnhancers) => {
      if (isTestingEnv) {
        return getDefaultEnhancers()
      }

      return getDefaultEnhancers().concat(devToolsEnhancer())
    },
  })

export const store = makeStore()
// we are going around a circular dependency here
setBackendStore(store)

export const persistor = persistStore(store)

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = typeof store.dispatch
