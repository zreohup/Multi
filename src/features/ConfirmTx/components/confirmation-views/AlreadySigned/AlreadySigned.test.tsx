import React from 'react'
import { render, fireEvent } from '@/src/tests/test-utils'
import { AlreadySigned } from './AlreadySigned'
import { Linking } from 'react-native'
import { GATEWAY_URL, SAFE_WEB_TRANSACTIONS_URL } from '@/src/config/constants'
import { http, HttpResponse } from 'msw'
import { server } from '@/src/tests/server'
import { apiSliceWithChainsConfig } from '@safe-global/store/gateway/chains'
import { makeStore } from '@/src/store'

// Mock Linking.openURL
const mockOpenURL = jest.fn()
Linking.openURL = mockOpenURL

describe('AlreadySigned', () => {
  const mockProps = {
    txId: 'test-tx-id',
    safeAddress: '0x123',
    chainId: '1',
  }

  const mockChain = {
    shortName: 'eth',
  }

  beforeEach(() => {
    jest.clearAllMocks()
    server.use(
      http.get(`${GATEWAY_URL}/v1/chains`, () => {
        return HttpResponse.json({
          results: [
            {
              chainId: '1',
              shortName: 'eth',
            },
          ],
        })
      }),
    )
  })

  const renderWithStore = async (ui: React.ReactElement) => {
    const store = makeStore()
    await store.dispatch(apiSliceWithChainsConfig.endpoints.getChainsConfig.initiate())
    return render(ui, { initialStore: store.getState() })
  }

  it('renders correctly with all required elements', async () => {
    const { getByText } = await renderWithStore(<AlreadySigned hasEnoughConfirmations={true} {...mockProps} />)

    expect(getByText('This transaction can be executed in the web app only.')).toBeTruthy()
    expect(getByText('Go to Web app')).toBeTruthy()
    expect(getByText('Confirm')).toBeTruthy()
  })

  it('opens web app URL when "Go to Web app" is pressed', async () => {
    const { getByText } = await renderWithStore(<AlreadySigned hasEnoughConfirmations={true} {...mockProps} />)

    const expectedUrl = SAFE_WEB_TRANSACTIONS_URL.replace(
      ':safeAddressWithChainPrefix',
      `${mockChain.shortName}:${mockProps.safeAddress}`,
    ).replace(':txId', mockProps.txId)

    fireEvent.press(getByText('Go to Web app'))
    expect(mockOpenURL).toHaveBeenCalledWith(expectedUrl)
  })

  it('renders correctly with all required elements', async () => {
    const { getByText } = await renderWithStore(<AlreadySigned hasEnoughConfirmations={false} {...mockProps} />)

    expect(getByText('Can be executed once the threshold is reached')).toBeTruthy()
  })

  it('matches snapshot', async () => {
    const { toJSON } = await renderWithStore(<AlreadySigned hasEnoughConfirmations={true} {...mockProps} />)
    expect(toJSON()).toMatchSnapshot()
  })
})
