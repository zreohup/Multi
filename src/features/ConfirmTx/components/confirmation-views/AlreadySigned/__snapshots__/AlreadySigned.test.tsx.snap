// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlreadySigned matches snapshot 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "column",
        "gap": 16,
        "justifyContent": "center",
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#A1A3A7",
          "fontFamily": "DM Sans",
          "fontSize": 14,
          "textAlign": "center",
        }
      }
      suppressHighlighting={true}
    >
      This transaction can be executed in the web app only.
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#121312",
              "fontFamily": "DMSans-Bold",
              "fontSize": 14,
              "textAlign": "center",
            }
          }
          suppressHighlighting={true}
        >
          Go to Web app
        </Text>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#121312",
                "fontSize": 16,
              },
              undefined,
              {
                "fontFamily": "SafeIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          î¤º
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "height": 50,
          "width": "100%",
        }
      }
    >
      <View
        focusVisibleStyle={
          {
            "borderColor": "$background",
          }
        }
        pointerEvents="none"
        role="button"
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#7878801F",
            "borderBottomColor": "transparent",
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderBottomWidth": 1,
            "borderLeftColor": "transparent",
            "borderLeftWidth": 1,
            "borderRightColor": "transparent",
            "borderRightWidth": 1,
            "borderStyle": "solid",
            "borderTopColor": "transparent",
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
            "borderTopWidth": 1,
            "bottom": 0,
            "color": "#636669",
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "fontWeight": 600,
            "height": "100%",
            "justifyContent": "center",
            "left": 0,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingBottom": 14,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 14,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="confirm-button"
      >
        <Text
          lineBreakMode="clip"
          numberOfLines={1}
          style={
            {
              "color": "#636669",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontFamily": "DMSans-SemiBold",
              "fontSize": 14,
              "letterSpacing": -0.1,
              "lineHeight": 15.400000000000002,
              "marginBottom": -2.5,
              "userSelect": "none",
            }
          }
          suppressHighlighting={true}
        >
          Confirm
        </Text>
      </View>
    </View>
  </View>
</View>
`;
